aten/src/ATen/native/Distance.cpp

```diff
@@ -88,8 +88,8 @@ Tensor cosine_similarity(const Tensor& x1, const Tensor& x2, int64_t dim, double
   Tensor w12 = at::sum(x1 * x2, dim);
   Tensor w1 = at::sum(x1 * x1, dim);
   Tensor w2 = at::sum(x2 * x2, dim);
-  Tensor n12 = (w1 * w2).rsqrt_().clamp_max(1.0 / eps);
-  return w12.mul_(n12);
+  Tensor n12 = (w1 * w2).sqrt_().clamp_min(eps);
```

Comment:
my guess is that you'll need to perform the clamp_min before the sqrt() to avoid NaNs in the derivative

Reply:
Yea I can do that, just double check we also need to update doc for it? As here https://github.com/pytorch/pytorch/blob/670f509984efce8aab0b0b5b166ff01472af6f1c/torch/nn/functional.py#L2838 we seem to imply we take `clamp_min` after `sqrt`.

Reply:
nvm, please ignore my naive question :P